### PR TITLE
feat(rspack_core): add oxc_resolver (turned off by default)

### PR DESCRIPTION
**Pull request by Boshen**
## Summary

This PR adds oxc_resolver to rspack_core. The resolver can be enabled by `experiments.newResolver`.

The next step is to switch the CI to use the new resolver, then eventually have it turned on by default.

I also plan to add `tracing` capabilities for debugging the resolver.

## Test Plan

* CI should pass because this is turned off by default
* I have tested the resolver on 3 internal projects and verified results against `nodejs_resolver`

### CODE REVIEW

File: crates/rspack_core/src/resolver/resolver_impl.rs [hunk 1]
```diff
@@ -68,12 +83,16 @@ impl<'a> ResolveInnerOptions<'a> {
...
 pub enum Resolver {
   NodejsResolver(nodejs_resolver::Resolver, Arc<nodejs_resolver::Cache>),
-  // OxcResolver(oxc_resolver::Resolver),
+  OxcResolver(oxc_resolver::Resolver),
 }
 
 impl Resolver {
   pub fn new(options: Resolve) -> Self {
-    Self::new_nodejs_resolver(options)
+    if std::env::var("OXC_RESOLVER").is_ok() {
```
Comment (hardfist) [inline crates/rspack_core/src/resolver/resolver_impl.rs:91]: use `experiments.rspackFuture.enableNewResolver` to control, so we can run  tests against new resolver and old resolver

Comment (h-a-n-a) [reply crates/rspack_core/src/resolver/resolver_impl.rs:91]: @hardfist  I thought we are omitting the `enable` part as we discussed before.

Comment (hardfist) [reply crates/rspack_core/src/resolver/resolver_impl.rs:91]: sorry forget this name convention

**Comment by hardfist** [comment]
can you help verify whether https://github.com/web-infra-dev/rspack/issues/2914 and https://github.com/web-infra-dev/nodejs_resolver/issues/185 is fixed in new resolver implementation

File: packages/rspack/src/config/defaults.ts [hunk 2]
```diff
@@ -157,6 +157,7 @@ const applyExperimentsDefaults = (
 	D(experiments, "lazyCompilation", false);
 	D(experiments, "asyncWebAssembly", false);
 	D(experiments, "newSplitChunks", true);
+	D(experiments, "newResolver", false);
```
Comment (hardfist) [inline packages/rspack/src/config/defaults.ts:160]: should be `experiments.rspackFuture.newResolver`

**Comment by Boshen** [comment]
> can you help verify whether https://github.com/web-infra-dev/rspack/issues/2914 and https://github.com/web-infra-dev/nodejs_resolver/issues/185 is fixed in new resolver implementation

The problem is fixed in the repro https://github.com/jmikrut/rspack-absolute-alias from the [mentioned comment](https://github.com/web-infra-dev/nodejs_resolver/issues/185#issuecomment-1653472531)

> https://github.com/web-infra-dev/rspack/tree/main/packages/rspack/tests/configCases/resolve these tests should be running against on newResolver and oldResolver

See https://github.com/web-infra-dev/rspack/pull/4108/commits/1e3f3b4848420d5bef6b32c722a305143635d9a3

**Comment by Boshen** [comment]
[1e3f3b4](https://github.com/web-infra-dev/rspack/pull/4108/commits/1e3f3b4848420d5bef6b32c722a305143635d9a3) Doesn't work, the tests will fail. What's t he proper way of doing this?

> https://github.com/web-infra-dev/rspack/tree/main/packages/rspack/tests/configCases/resolve these tests should be running against on newResolver and oldResolver

**Comment by Boshen** [comment]
> it would be helpful if you paste the performance result of newResolver and oldResolver in tested real world projects

I can't get a reliable benchmark from real projects due to too much noise from other parts of the compiler. The benchmark from the oxc should be good enough.

**Comment by hardfist** [comment]
!bench

**Comment by Boshen** [comment]
> !bench

Bench is not going to work because this is turned off by default.

File: packages/rspack/tests/ConfigCase.template.ts [hunk 3]
```diff
@@ -121,6 +121,12 @@ export const describeCases = config => {
 								if (!options.target) options.target = "node";
 								if (!options.output) options.output = {};
 								if (!options.output.path) options.output.path = outputDirectory;
+								if (config.newResolver) {
```
Comment (hardfist) [inline packages/rspack/tests/ConfigCase.template.ts:124]: shouldn't change this configuration, you could just copy configCase/resolve to configCase/new_resolve and add experiments.rspackFuture.newResolver to the webpack.config.js in each case https://github.com/web-infra-dev/rspack/blob/main/packages/rspack/tests/configCases/resolve/fallback/webpack.config.js#L10

**Comment by hardfist** [comment]
> > !bench
> 
> Bench is not going to work because this is turned off by default.

we may need to support cross compare performance of different configuration like webpack in the future, cc @jerrykingxyz

**Comment by jerrykingxyz** [comment]
> > > !bench
> > 
> > 
> > Bench is not going to work because this is turned off by default.
> 
> we may need to support cross compare performance of different configuration like webpack in the future, cc @jerrykingxyz

Yes, this is already planned.

File: crates/rspack_core/src/options/experiments.rs [hunk 4]
```diff
@@ -23,7 +23,9 @@ impl IncrementalRebuildMakeState {
 
 #[derive(Debug, Default)]
 #[allow(clippy::empty_structs_with_brackets)]
```
Comment (h-a-n-a) [inline crates/rspack_core/src/options/experiments.rs:25]:
```suggestion
```

Comment (h-a-n-a) [reply crates/rspack_core/src/options/experiments.rs:25]: Let's remove this, since it's useless now

**Review by hardfist — APPROVED** [review]

**Comment by hardfist** [comment]
@Boshen don't forget writing documentation about this feature

**Comment by hardfist** [comment]
@Boshen it seems oxc_resolver still missing missing.test.js and dependencies.test.js, which seems important to track dependency info during resolve process, will you implement it in the future?

**Pull request merged**